src/test/java/com/wikia/webdriver/pageobjectsfactory/componentobject/ad/VUAP.java

```diff
@@ -6,61 +6,67 @@
 import org.openqa.selenium.WebElement;
 
 public class VUAP {
-    private static By playTriggerButtonSelector = By.id("button");
-    private static By UIElementsSelector = By.className("overVideoLayer");
-    private final Wait wait;
-    private WikiaWebDriver driver;
-    private WebElement iframe;
-    private WebElement playTriggerButton;
+  private static final int VIDEO_LENGTH = 6000;
+  private static By videoContainerSelector = By.cssSelector(".video-ima-container.hidden");
+  private static By playTriggerButtonSelector = By.id("button");
+  private static By UIElementsSelector = By.className("overVideoLayer");
+  private final Wait wait;
+  private WikiaWebDriver driver;
+  private WebElement iframe;
+  private WebElement playTriggerButton;
 
-    public VUAP(WikiaWebDriver driver, String slotName) {
-        this.wait = new Wait(driver);
-        this.driver = driver;
-        iframe = findIframe(slotName);
-        playTriggerButton = findTriggerButton(driver);
-    }
+  public VUAP(WikiaWebDriver driver, String slotName) {
+    this.wait = new Wait(driver);
+    this.driver = driver;
+    setIframe(slotName);
+    setTriggerButton(driver);
+  }
 
-    private WebElement findIframe(String slotName) {
-        By iframeSelector = By.id(getAdUnit(slotName));
-        wait.forElementPresent(iframeSelector);
-        return driver.findElement(iframeSelector);
-    }
+  private String getAdUnit(String slotName) {
+    return "google_ads_iframe_/5441/wka.life/_project43//article/gpt/" + slotName + "_0";
+  }
 
-    private String getAdUnit(String slotName) {
-        return "google_ads_iframe_/5441/wka.life/_project43//article/gpt/" + slotName + "_0";
-    }
+  private void setIframe(String slotName) {
+    By iframeSelector = By.id(getAdUnit(slotName));
+    wait.forElementPresent(iframeSelector);
+    iframe = driver.findElement(iframeSelector);
+  }
 
-    private WebElement findTriggerButton(WikiaWebDriver driver) {
-        driver.switchTo().frame(iframe);
+  private void setTriggerButton(WikiaWebDriver driver) {
+    runInAdFrame(() -> {
+      wait.forElementClickable(playTriggerButtonSelector);
+      playTriggerButton = driver.findElement(playTriggerButtonSelector);
+    });
+  }
 
-        wait.forElementClickable(playTriggerButtonSelector);
-        WebElement element = driver.findElement(playTriggerButtonSelector);
-        driver.switchTo().defaultContent();
-        return element;
-    }
+  public WebElement getIframe() {
+    return iframe;
+  }
 
-    public WebElement getIframe() {
-        return iframe;
-    }
+  public void play() {
+    waitForVideoReadyToPlay();
+    runInAdFrame(() -> playTriggerButton.click());
+  }
 
-    public void play() {
-        driver.switchTo().frame(iframe);
+  private void runInAdFrame(Runnable f) {
+    driver.switchTo().frame(iframe);
+    f.run();
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) Call the method Thread.start() to execute the content of the run() method in a dedicated thread. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://sonar-s1/coding_rules#rule_key=squid%3AS1217)